### PR TITLE
Cancellable speech

### PR DESCRIPTION
**Pull request by feerrenrut**
Depends on refactoring in PR #11049, this diff will become easier to follow once that is merged.
Note: this fix is an experimental approach that must be turned on via the 'advanced settings panel' in the NVDA settings dialog.

### Summary of the issue:
In some web applications (such as Gmail) where user interaction causes rapid focus events NVDA may speak outdated focus information. This can be reproduced in Gmail by quickly moving through email  (in focus mode) using the down arrow. Sometimes two or three rows are reported from Gmail.

This issue happens more often in Chrome than in Firefox, but it can be reproduced in both browsers.

### Description of how this pull request fixes the issue:
This PR introduces a cancellable speech command, which can be inserted into a speech sequence for speech that may be invalidated (such as due to a new focus event)

### Testing performed:
Tested moving through emails (and across table boundaries) in Gmail.
Tested smoke tested web browsing.

### Change log entry:

Changes:
```
 - Web applications (EG Gmail) no longer speak outdated information when moving focus rapidly
```

### CODE REVIEW

**Comment by josephsl** [comment]
CC @LeonardDer as it can be generalized to include UIA notification handling.

File: source/NVDAObjects/IAccessible/ia2Web.py [hunk 1]
```diff
@@ -68,6 +68,16 @@ def _get_states(self):
 			states.discard(controlTypes.STATE_EDITABLE)
 		return states
 
+	def getFocusInfo(self):
```
Comment (leonardder) [inline source/NVDAObjects/IAccessible/ia2Web.py]: May be this can go on the IAccessible base class?

File: source/NVDAObjects/IAccessible/ia2Web.py [hunk 2]
```diff
@@ -68,6 +68,16 @@ def _get_states(self):
...
 		return states
 
+	def getFocusInfo(self):
+		return {
+			'accFocus': self.IAccessibleObject.accFocus,
+			'stateFocused': controlTypes.STATE_FOCUSED in self.states,
+			'indirectionsToAncestor': self._getIndirectionsToParentWithFocus(),
+		}
+
+	def isGainFocusValid(self):
```
Comment (leonardder) [inline source/NVDAObjects/IAccessible/ia2Web.py]: Same here, may be even on NVDAObject?

File: source/speech/__init__.py [hunk 3]
```diff
@@ -458,6 +474,34 @@ def _flattenNestedSequences(nestedSequences: Iterator[SpeechSequence]) -> Iterat
 	return [i for seq in nestedSequences for i in seq]
 
 
+def _getCancellableEvent(
```
Comment (leonardder) [inline source/speech/__init__.py]: It feels a bit odd that this is in the speech module, checking focus and such. May be some of this logic can be moved to eventHandler, so speech only has to query eventHandler and doesn't have to import NVDAObject here?

File: source/speech/__init__.py [hunk 4]
```diff
@@ -689,7 +733,7 @@ def getIndentationSpeech(indentation: str, formatConfig: Dict[str, bool]) -> Spe
 def speak(  # noqa: C901
 		speechSequence: SpeechSequence,
 		symbolLevel: Optional[int] = None,
-		priority: Optional[Spri] = None
+		priority: Spri = Spri.NORMAL
```
Comment (leonardder) [inline source/speech/__init__.py]: This looks unrelated.

**Review by LeonarddeR — COMMENTED** [review]
I did a small walk through this. At first glance, the cancelable speech is indeed pretty related to focus, whereas it could be more generic may be.

File: source/NVDAObjects/IAccessible/ia2Web.py (hunk 1 above)
Comment (feerrenrut) [reply source/NVDAObjects/IAccessible/ia2Web.py]: I'm hesitant to move this down if it's not used. Initially I'd prefer to focus on solving the specific problem rather than trying to generalize it.

File: source/config/configSpec.py [hunk 5]
```diff
@@ -249,6 +249,10 @@
 
 [development]
 	enableScratchpadDir = boolean(default=false)
+
+[featureFlag]
```
Comment (codeofdusk) [inline source/config/configSpec.py]: If we're creating a new config spec section for feature flags, do we want to move the existing ones there? For consistency maybe this should be part of speech settings in the config (but advanced in the GUI)?

File: source/config/configSpec.py [hunk 6]
```diff
@@ -249,6 +249,10 @@
 
 [development]
 	enableScratchpadDir = boolean(default=false)
+
+[featureFlag]
+	# 0:default, 1:yes, 2:no
+	cancelExpiredFocusSpeech = integer(0, 2, default=0)
```
Comment (codeofdusk) [inline source/config/configSpec.py]: Why not make this an option like winConsoleUIA?

File: source/config/configSpec.py (hunk 5 above)
Comment (feerrenrut) [reply source/config/configSpec.py]: In the case of this PR there is no intention of this becoming a relevant config option for most users. There is no plan to move this option into the speech section. After wider testing the default will change, and after some time has passed the GUI control can be removed assuming there are no problems. For stable GUI controls I think there is value in the config spec section matching the category. Feature flags are a little different, it helps us to introduce a change in behavior in a controlled way, allowing users to easily revert to prior behavior. I don't think there is much tangible benefit to moving preexisting feature flags to this section.

File: source/config/configSpec.py (hunk 6 above)
Comment (feerrenrut) [reply source/config/configSpec.py]: So that I don't have to have to map strings to translated strings. The value matches the index of the translated string in a list. This simplifies the GUI creation.

File: source/speech/__init__.py (hunk 4 above)
Comment (feerrenrut) [reply source/speech/__init__.py]: It is slightly, but an intentional change. I found myself trying to work out what the default priority was, and found I could simplify / clarify this.

**Comment by feerrenrut** [comment]
Just added a note to the description of this PR:
> When reviewing this, several of the first commits are refactoring. It will be much easier to review if you limit the commit range.

File: source/NVDAObjects/IAccessible/ia2Web.py (hunk 1 above)
Comment (michaelDCurran) [inline source/NVDAObjects/IAccessible/ia2Web.py:71]: I can't see where this method is actually called? perhaps it is not needed anymore?

File: source/NVDAObjects/IAccessible/ia2Web.py (hunk 2 above)
Comment (michaelDCurran) [inline source/NVDAObjects/IAccessible/ia2Web.py:78]: I also don't think this is used anymore?

**Review by michaelDCurran — CHANGES_REQUESTED** [review]
Firstly this diff was very hard to read with all the refactoring going on. True I could look at commits individually, but I do feel this sort of suggests that refactoring would have been better in a separate pr first if it were truly necessary.
As for the actual changes needed in this pr:  if I understand correctly, in simple terms, if a speech utterance was produced due to a gainFocus event, it will contain a cancellableSpeechCommand. This cancellableSpeechCommand will show itself as 'canceled' if the object the event was for was definitly queued as a focus event at some point, and now is not the currently queued focus event. Is my understanding correct?
There are some extra changes to do with focus on this pr, such as the addition of focusInfo and isGainFocusValid. But I cannot see where these are actually ever called... unless NVDA / github are not playing nicely for me...
If they are not called, then they should be obviously removed.
I can confirm that this approach does solve the issue in gmail.

**Comment by feerrenrut** [comment]
@michaelDCurran Your summary sounds correct to me. I'll clean up those unused functions and split the PR into a refactor + feature. Thanks for reviewing it.

**Comment by feerrenrut** [comment]
Refactor for cancellable speech #11049 has now been merged and this PR rebased. It is now ready to be reviewed again.

**Comment by Adriani90** [comment]
I think this will probably fix #6421

**Review by michaelDCurran — DISMISSED** [review]
I really don't think changes to ```NVDAObjects\IAccessible\__init__.py``` and ```NVDAObjects\IAccessible\Ia2web.py``` are necessary in this pr. 
However, I am happy with the rest of it.

**Comment by codeofdusk** [comment]
With this PR merged, some lines are repeating in say all in Chrome 83.0.4103.44 even with the feature flag disabled.

**Comment by feerrenrut** [comment]
Can you create a new issue with steps to reproduce? I tried Chrome and Chrome Canary with the feature flag disabled using say-all on wikipeda's "from today's featured article" and couldn't reproduce the problem. Please test with all add-ons disabled to confirm.

**Comment by codeofdusk** [comment]
Tested with all add-ons disabled.

Read the description of this PR. As one example, the sentence "Tested moving through emails (and across table boundaries) in Gmail" is read twice.

**Comment by codeofdusk** [comment]
Start the say all while focused on the "Description of how this pull request fixes the issue" heading.

**Comment by feerrenrut** [comment]
Ok, I can reproduce that issue with `alpha-20175,9967903a` and it does not occur in `alpha-20168,1cf36fbe`

**Comment by feerrenrut** [comment]
I'll look at fixing this early next week. If we can't fix this issue before branching for beta the PR will have to be reverted.

**Comment by feerrenrut** [comment]
Created issue:  #11132

I'll look at fixing this early next week. If we can't fix this issue before branching for beta the PR will have to be reverted.

**Comment by Adriani90** [comment]
My understanding is that this pr targets an advanced setting, so it is an experimental setting. I think there is no need to revert since the users know that the advanced settings can possibly contain preliminary work. What do you think?

Von meinem iPhone gesendet

**Comment by feerrenrut** [comment]
@Adriani90 you might have missed it, this problem also occurs when the advanced setting is disabled. The code in speech manager (particularly how it interacts with say-all) is fragile, it's difficult to make modifications and easy make mistakes.

**Comment by Adriani90** [comment]
@feerrenrut before creating a new issue, I just wanted to share that enabling this option causes NVDA to skip reporting of window title or tab title when switching with alt+tab or with ctrl+tab. Maybe it is possible to limit the cancelable speech to the elements of the window except for address bars, status bars, notification bars and window titles.
If this does not get reverted next week, I will create a new issue for this specific problem.

**Comment by feerrenrut** [comment]
Note: #11144

**Comment by ramchaitanyap** [comment]
Lets say, I change focus by some timeout, NVDA is still not cancelling expired focus speech.

function getFocus() {
document.getElementById("myAnchor").focus();
setTimeout(() => document.getElementById("myAnchor2").focus(), 500);
}

Its still reading #myAnchor text and then #myAnchor2 text
